### ISSUE TITLE
IRCClient.say passes None to IRCClient.msg, which causes an exception to be raised.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4758]
|<img alt="Bobng's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Bobng reported|
|-|-|
|Trac ID|trac#4758|
|Type|release blocker: regression|
|Created|2010-12-04 23:43:27Z|
|Branch|https://github.com/twisted/twisted/tree/ircclient-say-4758|

When using IRCClient.say, the default value for "length" is None. This then gets passed to IRCClient.msg, which then raises a very unhelpful exception like the one below:
exceptions.ValueError: Maximum length must exceed [Length] for message to [Channel].

The fix is very simple: Make IRCClient.say's "length" kwarg use MAX_COMMAND_LENGTH as its default value, and/or make IRCClient.msg use MAX_COMMAND_LENGTH if an integer is not given in the "length" kwarg.

I rated this "high" because any previous code using IRCClient.say without explicitly passing a "length" attribute will fail.

<details><summary>Searchable metadata</summary>

```
trac-id__4758 4758
type__release_blocker__regression release blocker: regression
reporter__Bobng Bobng
priority__high high
milestone__ 
branch__branches_ircclient_say_4758 branches/ircclient-say-4758
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1291506207000000 1291506207000000
changetime__1294974621000000 1294974621000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

I think this is a defect, not a regression, unless someone would like to point to where it's introduced...

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

It seems quite likely that [#4416](https://github.com/twisted/twisted/issues/4416) introduced this problem, by changing the default value of the `length` parameter to `IRCClient.msg` from `None` to `MAX_COMMAND_LENGTH`.  Now any callers which pass `None` explicitly will cause a comparison between `None` and the minimum possible length limit, which will always cause the `msg` call to fail.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [30365]) Branching to 'ircclient-say-4758'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [30366]) A test for IRCClient.say (the first) and a test for passing None as the length to IRCClient.msg.  And a fix for both.


refs [#4758](https://github.com/twisted/twisted/issues/4758)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ircclient-say-4758)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

[#4788](https://github.com/twisted/twisted/issues/4788) was a duplicate of this.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli set owner to @exarkun|
|-|-|

Please bump the copyright headers on both files, otherwise looks good to merge!  :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [30438]) news fragment

refs [#4758](https://github.com/twisted/twisted/issues/4758)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [30439]) Bump copyright date

refs [#4758](https://github.com/twisted/twisted/issues/4758)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [30440]) Merge ircclient-say-4758

Author: exarkun
Reviewer: cyli
Fixes: [#4758](https://github.com/twisted/twisted/issues/4758)

Make a `None` value for the `length` parameter to `IRCClient.msg` the same as
passing `MAX_COMMAND_LENGTH` as the value for that parameter.  This restores
pre-10.2 behavior for both `IRCClient.msg` and `IRCClient.say`.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
